event/event.go

```diff
@@ -50,3 +51,69 @@ func (p *Publisher[T]) Publish(ctx context.Context, event T) error {
 		Body: encBody,
 	})
 }
+
+// RawSubscription represents a subscription that delivers messages as is.
+// No assumptions are made about the message contents. This should rarely be used in favor of [Subscription].
+type RawSubscription struct {
+	sub            *pubsub.Subscription
+	maxConcurrency int
+}
+
+// RawMessageHandler is responsible for handling raw messages from a subscription.
+type RawMessageHandler func([]byte) error
+
+// NewRawSubscription creates a new raw subscription. It provides messages in a
+// service like manner (serve) and manages concurrent execution, each message
+// is processed in its own goroutines respecting the given maxConcurrency.
+func NewRawSubscription(url string, maxConcurrency int) (*RawSubscription, error) {
+	if maxConcurrency <= 0 {
+		return nil, fmt.Errorf("max concurrency must be > 0: %d", maxConcurrency)
+	}
+	// We dont want the subscription to expire, so we use the background context.
+	sub, err := pubsub.OpenSubscription(context.Background(), url)
+	if err != nil {
+		return nil, err
+	}
+	return &RawSubscription{
+		sub:            sub,
+		maxConcurrency: maxConcurrency,
+	}, nil
+}
+
+// Serve will start serving all messages from the subscription calling handler for each
+// message. It will run until [RawSubscription.Shutdown] is called.
+// If the error is nil Ack is sent.
+// If a non-nil error is returned by the handler Unack will be sent.
+// Serve may be called multiple times, each time will start a new serving service that will
+// run up to "maxConcurrency" goroutines.
+func (r *RawSubscription) Serve(handler RawMessageHandler) error {
+	semaphore := make(chan struct{}, r.maxConcurrency)
+	for {
+		semaphore <- struct{}{}
+		msg, err := r.sub.Receive(context.Background())
+		if err != nil {
+			// From: https://pkg.go.dev/gocloud.dev@v0.30.0/pubsub#example-Subscription.Receive-Concurrent
+			// Errors from Receive indicate that Receive will no longer succeed.
+			return fmt.Errorf("receive from subscription failed, stopping serving: %v", err)
+		}
+		go func() {
+			defer func() {
+				<-semaphore
+			}()
+			err := handler(msg.Body)
+			if err != nil {
+				if msg.Nackable() {
+					msg.Nack()
+				}
+				return
+			}
+			msg.Ack()
+		}()
+	}
+}
+
+// Shutdown will shutdown the subscriber, stopping any calls to [RawSubscription.Serve].
+// The subscription should not be used after this method is called.
+func (r *RawSubscription) Shutdown(ctx context.Context) error {
+	return r.sub.Shutdown(ctx)
```

Comment:
Our use-cases is slighly different then the example, but don't we need to wait for all messages that are being handling? Or shutdown would solve that?

Reply:
Shutdown will flush all pending acks: https://pkg.go.dev/gocloud.dev@v0.30.0/pubsub#Subscription.Shutdown

But it won't wait for all outstanding messages to be handled. We could implement this behavior here but Im not sure we need it (messages are just sent back to the subscription and retried on a different process, unless this happens all the time it should not cause major issues).

Reply:
Mmh, this wouldn't be a problem in an environment where we have idempotency, which, AFAIK it's not our case. we can call multiple times the same API, publish the same event multiple time,s and so on.

This can happen anyway, but if we don't handle it feels like we're increasing the probabilities as any new deployment we're adding the change of happening increase.

We can treat it as an improvement, though, I just feel should be something that we should have before we integrate it everywhere.

Reply:
which events we have that if processed twice causes issues ? :thinking: AFAIK it would only generate unnecessary reprocessing in some cases..but no severe issues (like debiting money twice etc). When reprocessing is an issue I prefer to have explicit idempotency mechanisms, trusting the broker to deliver only once is tricky IMHO.

For example, google has it: https://cloud.google.com/pubsub/docs/exactly-once-delivery
And yet depending on what happens on the handlers, it may be delivered to more than one (after the ack timeout), so if you are doing something that could cause problems if done twice you will need to handle it on the code anyway..or have severe problems.

+ If this is as essential as you say it doesn't change the interface of shutdown..so adding it later and updating the libraries should be fairly OK too. So not so sure we need this right away :thinking:.

Reply:
Yep, I'm aware it can happen anyway. I just feel that not handling it increases the probability of happening on every deployment. But I also acknowledge that it can be treated as improvement, so all good!

Reply:
To be on the safe side we can add this right away, can do in a separate PR then.